Python/Source/ListTrial.cpp

```diff
@@ -2,24 +2,23 @@
 #include "Sample.hpp"
 
 using namespace pybind11;
+using namespace Rover;
 
-namespace Rover {
-  void export_list_trial(pybind11::module& module) {
-    class_<ListTrial<PythonSample>>(module, "ListTrial")
-      .def(init<>())
-      .def("reserve", &ListTrial<PythonSample>::reserve)
-      .def("capacity", &ListTrial<PythonSample>::capacity)
-      .def("insert", (void (ListTrial<PythonSample>::*)(const PythonSample&))
-         &ListTrial<PythonSample>::insert)
-      .def("insert", [](ListTrial<PythonSample>& out,
-           ListTrial<PythonSample> in) {
-         out.insert(std::make_move_iterator(in.begin()),
-           std::make_move_iterator(in.end()));
-       })
-      .def("__getitem__", &ListTrial<PythonSample>::operator [])
-      .def("__iter__", [](const ListTrial<PythonSample>& t) {
-         return make_iterator(t.begin(), t.end());
-       })
-      .def("__len__", &ListTrial<PythonSample>::size);
-  }
+void Rover::export_list_trial(pybind11::module& module) {
+  class_<ListTrial<PythonSample>>(module, "ListTrial")
+    .def(init<>())
+    .def("reserve", &ListTrial<PythonSample>::reserve)
+    .def("capacity", &ListTrial<PythonSample>::capacity)
+    .def("insert", static_cast<void (ListTrial<PythonSample>::*)(const
+       PythonSample&)>(&ListTrial<PythonSample>::insert))
+    .def("insert", [](ListTrial<PythonSample>& out,
+         ListTrial<PythonSample> in) {
+       out.insert(std::make_move_iterator(in.begin()),
```

Comment:
After changing ```in``` to a ```const&```, you won't need to make a move iterator.

Reply:
Isn't the copy-and-move idiom applicable here?

Reply:
It's a good point but in this case you're not moving the list itself, only its elements. Making a copy of the list is the expensive operation relatively speaking and that list doesn't get moved/reused so the copy will go to waste. On the other hand making a copy of the elements is relatively cheap.

Reply:
So shall I only do copy-and-move when I can move class data members directly? Or is it on a case-by-case basis, e.g. here we assume that vectors are large compared to individual Samples and their potential copy overhead?
I am somewhat opposed to copying large structures myself as that, at very least, will double the peak memory usage, but I am just trying to understand the convention here.

Reply:
Yes that's fair, the copy-and-move idiom is only applicable when invoking a move constructor, it doesn't apply to assignment operators. Example:

```
// Copy-move idiom not applicable as an optimization since this is an assignment.
// Compilers don't perform move-elision for assignment. Pass by const& or add an r-value overload.
void f(const std::string& s) {
  this.s = s;
}

// Copy move idiom is applicable since this will invoke the move-constructor.
void f(std::string s) {
  this.some_list.push_back(std::move(s));
}

// Copy-move idiom is not applicable here since the vector itself isn't getting moved but rather only its elements which are very cheap to copy relative to the cost of copying a vector.
void f(const std::vector<int>& s) {
  this.some_list.insert(this.some_list.end(), s.begin(), s.end());
}
```